File: src/face_id_kit/live.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field
import time
from .matching import FaceIdentifier
from .types import LiveResult


def iou(left, right):
    lx, ly, lw, lh = left
    rx, ry, rw, rh = right
    overlap = max(0., min(lx + lw, rx + rw) - max(lx, rx)) * max(0., min(ly + lh, ry + rh) - max(ly, ry))
    union = lw * lh + rw * rh - overlap
    return overlap / union if union > 0 else 0.


@dataclass
class _Track:
    bbox: tuple[float, float, float, float]
    seen_at: float
    history: deque = field(default_factory=deque)


class TrackConsensus:
    """Conservative geometric association. Ambiguous overlaps start fresh tracks."""
    def __init__(self, *, window: int = 6, required: int = 4,
                 expiry_seconds: float = 1.0, min_iou: float = 0.3):
        if not 1 <= required <= window or expiry_seconds <= 0 or not 0 < min_iou <= 1:
            raise ValueError("Invalid tracking configuration")
        self.window, self.required = window, required
        self.expiry_seconds, self.min_iou = expiry_seconds, min_iou
        self._tracks: dict[int, _Track] = {}
        self._next_id = 0

    def reset(self):
        self._tracks.clear()

    def update(self, boxes, identity_ids, *, now: float | None = None):
        if len(boxes) != len(identity_ids):
            raise ValueError("Boxes and identities must have the same length")
        now = time.monotonic() if now is None else now
        self._tracks = {k: v for k, v in self._tracks.items() if now - v.seen_at < self.expiry_seconds}
        options = [[tid for tid, track in self._tracks.items() if iou(box, track.bbox) >= self.min_iou] for box in boxes]
        counts = {tid: sum(tid in choices for choices in options) for tid in self._tracks}
        ambiguous = {tid for choices in options for tid in choices if len(choices) != 1 or counts[tid] != 1}
        for tid in ambiguous:
            del self._tracks[tid]
        result, used = [], set()
        for box, identity, choices in zip(boxes, identity_ids, options):
            if len(choices) == 1 and choices[0] in self._tracks and counts[choices[0]] == 1:
                tid = choices[0]
                track = self._tracks[tid]
            else:
                tid = self._next_id
                self._next_id += 1
                track = _Track(tuple(box), now, deque(maxlen=self.window))
                self._tracks[tid] = track
            if track.history and identity is not None and track.history[-1] not in (identity, None):
                track.history.clear()
            track.bbox, track.seen_at = tuple(box), now
            track.history.append(identity)
            used.add(tid)
            result.append((tid, identity is not None and track.history.count(identity) >= self.required))
        for tid, track in self._tracks.items():
            if tid not in used:
                track.history.append(None)
        return result


class LiveSession:
    def __init__(self, faces: FaceIdentifier, *, max_faces: int | None = 4,
                 central_only: bool = False, tracker: TrackConsensus | None = None):
        self.faces, self.max_faces, self.central_only = faces, max_faces, central_only
        self.tracker = tracker or TrackConsensus()
        self._revision = faces.store.revision

    def update(self, image_bgr):
        revision = self.faces.store.revision
        if revision != self._revision:
            self.tracker.reset()
            self._revision = revision
        results = self.faces.identify(image_bgr, max_faces=self.max_faces, central_only=self.central_only)
        tracks = self.tracker.update([f.bbox for f, _ in results], [m.identity_id for _, m in results])
        return [LiveResult(tid, face, match, stable) for (face, match), (tid, stable) in zip(results, tracks)]

File: android/core/src/main/kotlin/io/github/evoltriet/faceidkit/Live.kt
package io.github.evoltriet.faceidkit

private data class Track(var box: Box, var seen: Double, val history: MutableList<String?> = mutableListOf())
class TrackConsensus(private val window: Int = 6, private val required: Int = 4,
    private val expirySeconds: Double = 1.0, private val minIou: Double = .3) {
    init { require(required in 1..window && expirySeconds > 0 && minIou > 0 && minIou <= 1) }
    private val tracks = linkedMapOf<Long, Track>()
    private var nextId = 0L
    fun reset() { tracks.clear() }
    private fun append(track: Track, id: String?) { track.history.add(id); while (track.history.size > window) track.history.removeAt(0) }
    fun update(boxes: List<Box>, identities: List<String?>, now: Double = System.nanoTime() / 1e9): List<Pair<Long, Boolean>> {
        require(boxes.size == identities.size)
        tracks.entries.removeAll { now - it.value.seen >= expirySeconds }
        val options = boxes.map { box -> tracks.filterValues { box.iou(it.box) >= minIou }.keys.toList() }
        val counts = tracks.keys.associateWith { id -> options.count { id in it } }
        val ambiguous = options.flatMap { choices -> choices.filter { choices.size != 1 || counts[it] != 1 } }.toSet()
        ambiguous.forEach { tracks.remove(it) }
        val used = mutableSetOf<Long>()
        val result = boxes.indices.map { i ->
            val choices = options[i]
            val id = if (choices.size == 1 && choices[0] in tracks && counts[choices[0]] == 1) choices[0] else nextId++
            val track = tracks.getOrPut(id) { Track(boxes[i], now) }
            val identity = identities[i]
            if (track.history.isNotEmpty() && identity != null && track.history.last() != null && track.history.last() != identity) track.history.clear()
            track.box = boxes[i]; track.seen = now; append(track, identity); used.add(id)
            id to (identity != null && track.history.count { it == identity } >= required)
        }
        tracks.filterKeys { it !in used }.values.forEach { append(it, null) }
        return result
    }
}
class LiveSession(private val faces: FaceIdentifier, private val maxFaces: Int = 4,
    private val centralOnly: Boolean = false, val tracker: TrackConsensus = TrackConsensus()) {
    private var revision = faces.store.revision
    @Synchronized fun reset() { tracker.reset() }
    @Synchronized fun update(image: BgrImage, now: Double = System.nanoTime() / 1e9): List<LiveResult> {
        val before = faces.store.revision
        if (before != revision) { tracker.reset(); revision = before }
        val results = faces.identify(image, maxFaces, centralOnly)
        if (before != faces.store.revision) { tracker.reset(); return emptyList() }
        val tracks = tracker.update(results.map { it.first.box }, results.map { it.second.identityId }, now)
        return results.zip(tracks).map { (r, track) -> LiveResult(track.first, r.first, r.second, track.second) }
    }
}
